usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +519,467 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+    pkey_wrap_handler_data.secure_key = skey_attr->pValue;
+    pkey_wrap_handler_data.secure_key_len = skey_attr->ulValueLen;
+    pkey_wrap_handler_data.pkey_buf = (CK_BYTE *)&ep11_buf;
+    pkey_wrap_handler_data.pkey_buflen_p = &ep11_buflen;
+    ret = handle_all_ep11_cards(&ep11_data->target_list, ep11tok_pkey_wrap_handler,
+                                &pkey_wrap_handler_data);
+    if (ret != CKR_OK || !pkey_wrap_handler_data.wrap_was_successful) {
+        TRACE_ERROR("handle_all_ep11_cards failed or no APQN could do the wrap.\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Check ep11 wrapped key struct length */
+    if (ep11_buflen != sizeof(wrapped_key_t)) {
+        TRACE_ERROR("invalid ep11 wrapped key struct length %ld\n", ep11_buflen);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Check if returned key type is what we expect:
+     *  - 0x1 = AES with bit length 128, 192 and 256
+     *  - 0x2 = 2DES and 3DES
+     *  - 0x3 = EC-P with bit length 192, 224, 256, 384 and 521
+     *  - 0x4 = ED25519 and ED448
+     */
+    wk = (wrapped_key_t *) &ep11_buf;
+    switch (wk->cpacf_key_type) {
+    case 1:
+    case 3:
+        break; // ok.
+    default:
+        TRACE_ERROR("Got unexpected CPACF key type %d from firmware\n", wk->cpacf_key_type);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Build new attribute for protected key */
+    ret = build_attribute(CKA_IBM_OPAQUE_PKEY, wk->wrapped_key,
+                          wk->token_size, &tmp_attr);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("build_attribute failed with rc=0x%lx\n", ret);
+        ret = CKR_FUNCTION_FAILED;;
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    *pkey_attr = tmp_attr;
+
+    return ret;
+}
+
+/**
+ * Save the current firmware master key verification pattern in the tokdata:
+ * create a dummy test key, transform it into a protected key, and store the MK
+ * verification pattern in the tokdata.
+ */
+static CK_RV ep11tok_pkey_get_firmware_mk_vp(STDLL_TokData_t *tokdata)
+{
+    CK_BBOOL btrue = CK_TRUE;
+    CK_ULONG len = AES_KEY_SIZE_256;
+    CK_MECHANISM mech = {CKM_AES_KEY_GEN, NULL_PTR, 0};
+    CK_ATTRIBUTE tmpl[] = {
+        {CKA_VALUE_LEN, &len, sizeof(CK_ULONG)},
+        {CKA_IBM_PROTKEY_EXTRACTABLE, &btrue, sizeof(btrue)},
+    };
+    CK_ULONG tmpl_len = sizeof(tmpl) / sizeof(CK_ATTRIBUTE);
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_BYTE csum[MAX_CSUMSIZE];
+    size_t csum_l = sizeof(csum);
+    CK_BYTE blob[MAX_BLOBSIZE];
+    size_t blobsize = sizeof(blob);
+    CK_ATTRIBUTE *pkey_attr = NULL, *blob_attr=NULL;
+    CK_RV ret;
+
+    /* Check if CPACF_WRAP mech supported */
+    if (ep11tok_is_mechanism_supported(tokdata, CKM_IBM_CPACF_WRAP) != CKR_OK) {
+        TRACE_INFO("CKM_IBM_CPACF_WRAP not supported on this system.\n");
+        ret = CKR_FUNCTION_NOT_SUPPORTED;
```

Comment:
Does this mean that the token does not initialize if pkey is enabled in the config, but the system does not support it? If so, then I would also issue a syslog message here. Otherwise the user won't have a chance to know why the token does not initialize.

Reply:
The syslog message is now issued after returning from ep11tok_pkey_get_firmware_mk_vp.

Reply:
Im not sure about LOG_WARNING. We had the discussion if libraries should write to the syslog at all multiple times. Idk what openCryptoki does in similar situations, however, id prefer if all syslogging would be at debug level.